server/ipqueue.go

```diff
@@ -101,81 +110,102 @@ func newIPQueue[T any](s *Server, name string, opts ...ipQueueOpt[T]) *ipQueue[T
 }
 
 // Add the element `e` to the queue, notifying the queue channel's `ch` if the
-// entry is the first to be added, and returns the length of the queue after
-// this element is added.
-func (q *ipQueue[T]) push(e T) (int, error) {
-	var signal bool
+// element is the first to be added, and returns the length and size of the queue
+// after this element is added.
+// If the queue has limits, this function checks against what is currently in
+// the queue in addition to the "in progress" elements (the ones that were
+// returned by `pop()`). It will return `errIPQLenLimitReached` or
+// `errIPQSizeLimitReached` if adding the element would have gone over the length
+// or size limit respectively.
+func (q *ipQueue[T]) push(e T) (int, uint64, error) {
+	var sz uint64
+	if q.calc != nil {
+		sz = q.calc(e)
+	}
 	q.Lock()
 	l := len(q.elts) - q.pos
-	if q.mlen > 0 && l == q.mlen {
+	qsz := q.sz
+	// If max length is specified, check the current length + "in progress"
+	// count to see if we are going to go over the limit.
+	if q.mlen > 0 && l+int(atomic.LoadInt64(&q.inprogress))+1 > q.mlen {
 		q.Unlock()
-		return l, errIPQLenLimitReached
+		return l, qsz, errIPQLenLimitReached
 	}
-	if l == 0 {
-		signal = true
-		eltsi := q.pool.Get()
-		if eltsi != nil {
-			// Reason we use pointer to slice instead of slice is explained
-			// here: https://staticcheck.io/docs/checks#SA6002
-			q.elts = (*(eltsi.(*[]T)))[:0]
-		}
-		if cap(q.elts) == 0 {
-			q.elts = make([]T, 0, 32)
-		}
-	}
-	if q.calc != nil {
-		sz := q.calc(e)
-		if q.msz > 0 && q.sz+sz > q.msz {
+	if sz > 0 {
+		// Take into account the "in progress" size.
```

Comment:
This is a good change and it may be worth breaking this out into its own PR.

Reply:
Ok, so this PR would just be about using *[]T all the way only? But then it means that the pop() would just return the *[]T and then the recycle() take that and update the in progress count and that's it. The next PR would introduce the in progress size and have pop() returns the *[]T, list Len, list size and basically be like it is in this PR.